### ISSUE TITLE
validate_certs option is not listed as required in docs and defaults to true

### ISSUE THREAD

**Issue by selhorn** [F5Networks/f5-ansible#563]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
bigip_node

##### ANSIBLE VERSION
```
ansible 2.4.2.0
  config file = None
  configured module search path = [u'/home/ubuntu/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python2.7/dist-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.12 (default, Nov 20 2017, 18:23:56) [GCC 5.4.0 20160609]
```

##### PYTHON VERSION
```
Python 2.7.12
```

##### BIGIP VERSION
```
Sys::Version
Main Package
  Product     BIG-IP
  Version     13.0.0
  Build       3.0.1679
  Edition     Hotfix HF3
  Date        Fri Nov 10 23:41:13 PST 2017

Hotfix List
ID641482-3    ID639767-1  ID642449     ID633091    ID635126-2  ID561596
ID572234-1    ID366695-9  ID635116-3   ID643326    ID628337-2  ID646511-2
ID642124      ID636663    ID639526     ID641963    ID598024-1  ID588414-1
ID635680-1    ID636044-2  ID638799-2   ID643328    ID641360-1  ID638779
ID641013-6    ID578076    ID636016     ID627695-1  ID609995-1  ID598002-9
ID644041      ID603609-1  ID629921-3   ID629752    ID650450    ID569100
ID640824-2    ID621976-5  ID563165     ID626594-3  ID638838    ID605792
ID638629-1    ID643752-1  ID654368-1   ID649617-1  ID607246-8  ID642613
ID645805-2    ID631715-2  ID598289     ID643327    ID643332    ID596924-1
ID637094      ID632504-2  ID429213     ID643396-1  ID639406    ID637181-1
ID651627      ID611161-4  ID630611-3   ID649048    ID649571-2  ID640829-1
ID643547-2    ID633391-2  ID652151     ID633879-2  ID649369-1  ID629013
ID636520-1    ID352957-2  ID643187-1   ID622160    ID649933-2  ID570855-1
ID419741-4    ID620759-3  ID639049-1   ID631572    ID639575-2  ID636675
ID641547-1    ID638825-1  ID578017     ID636853-4  ID639565    ID508113-2
ID630610-1    ID636700-1  ID645203-1   ID637666-1  ID590091-4  ID633564-1
ID628164-4    ID575642    ID647706-1   ID636699-6  ID541320-8  ID640510-2
ID642562      ID637847    ID517756-5   ID648056-3  ID639764-1  ID629017
ID641390-2    ID647962-1  ID609200-1   ID617901-9  ID639383    ID640352-1
ID636702-4    ID630390-1  ID572272     ID642400-3  ID642119-1  ID640854
ID651476-1    ID638576-1  ID247527-1   ID550547-1  ID664063    ID559080
ID610122      ID627554-1  ID620659-4   ID636790-4  ID642983-2  ID658764
ID640457-3    ID644855-1  ID648867-1   ID644817-1  ID552988-1  ID489499-2
ID636613      ID630712    ID621974-5   ID637141-1  ID611691-6  ID608245-1
ID594228-1    ID636479    ID623362-1   ID635111-1  ID638219    ID643325
ID610485-1    ID641612-1  ID642926-1   ID641724    ID645219    ID631688-8
ID646719-1    ID637308-1  ID582773     ID608304-2  ID367226-3  ID639774-1
ID641445-2    ID638780    ID620625-3   ID619593-1  ID615372    ID626386-2
ID600205-1    ID636155    ID541550-2   ID644490-2  ID635754-2  ID639528
ID648786-1    ID639744-3  ID610201-1   ID633985-1  ID662108    ID636289-1
ID642090      ID640376-2  ID615267     ID639530    ID627747    ID581746-6
ID642015-4    ID642058    ID629085-2   ID637664-1  ID638495-1  ID644805
ID639630-1    ID639288-1  ID643404-1   ID618332-3  ID638215    ID635189-1
ID643330      ID638115-1  ID606710-11  ID641083-1  ID618430-1  ID632499-2
ID569814      ID652959-1  ID638967-2   ID642211-1  ID632668-6  ID638556-1
ID639729-1    ID639500-1  ID648990-1   ID660170-2  ID561592    ID642330
ID624896-1    ID570926    ID625892-1   ID610307-4  ID621379-1  ID637561-2
ID632060-2    ID648715-3  ID635688     ID633723-4  ID641169    ID642874-2
ID635933      ID626861-1  ID642221-1   ID624722    ID693211-2  ID651106-1
ID417720      ID644870    ID639486-1

~
~
~
~

```

##### LIBRARY VERSIONS
```
bigsuds==1.0.6
f5-sdk==3.0.7
```

##### CONFIGURATION
none

##### OS / ENVIRONMENT
Ubuntu 16.04 LTS 20171121.1 in AWS

##### SUMMARY
These three modules produce an error if you do not include the validate_certs option. 
bigip_node
bigip_pool_member
bigip_virtual_server

(These are the ones I tested, there may be others) (bigip_pool has this option but does not have this issue)

The message is:
fatal: [50.18.101.9]: FAILED! => {"changed": false, "msg": "received exception: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:590)>"}

More details below. It seems to assume **true** if you do not include this option.

The docs say that this option is **not required**. So either the docs or the module should be updated.

If you re-run with validate_certs: "false", it works fine.

##### STEPS TO REPRODUCE

```
IP addresses, username/password changed for sake of bug. Here is playbook with one of the modules:

---

- name: Create a VIP, pool, pool members and nodes
  hosts: 10.10.10.10
  connection: local
  
  tasks:
      - name: Create a VIP
        bigip_virtual_server:
            server: "10.10.10.10"
            user: "blah"
            password: "blah"
            name: "vip-1"
            description: "vip1"
            destination: "10.10.10.20"
            port: "80"
            snat: "Automap"
            all_profiles:
                - "http"
                - "clientssl"
            pool: "web_pool"

And then the same playbook was used, with this added, to get a successful response:
validate_certs: "false"

```

##### EXPECTED RESULTS
Should assume false if the validate_certs option is not specified, or the docs should say this option is required. Not sure.

##### ACTUAL RESULTS

```
The full traceback is:
  File "/tmp/ansible_qI41MA/ansible_module_bigip_virtual_server.py", line 776, i                                                                                                                                                             n main
    if not vs_exists(api, name):
  File "/tmp/ansible_qI41MA/ansible_module_bigip_virtual_server.py", line 217, i                                                                                                                                                             n vs_exists
    api.LocalLB.VirtualServer.get_object_status(virtual_servers=[vs])
  File "/usr/local/lib/python2.7/dist-packages/bigsuds.py", line 379, in __getat                                                                                                                                                             tr__
    client = self._client_creator('%s.%s' % (self._name, attr))
  File "/usr/local/lib/python2.7/dist-packages/bigsuds.py", line 189, in _create                                                                                                                                                             _client
    raise ConnectionError(str(e))

fatal: [10.10.10.10]: FAILED! => {
    "changed": false,
    "invocation": {
        "module_args": {
            "all_policies": null,
            "all_profiles": [
                "http",
                "clientssl"
            ],
            "all_rules": null,
            "default_persistence_profile": null,
            "description": "vip1",
            "destination": "10.0.1.202",
            "enabled_vlans": null,
            "fallback_persistence_profile": null,
            "name": "vip-1",
            "partition": "Common",
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "pool": "web_pool",
            "port": "80",
            "route_advertisement_state": null,
            "server": "10.10.10.10",
            "server_port": 443,
            "snat": "Automap",
            "state": "present",
            "user": "admin",
            "validate_certs": true
        }
    },
    "msg": "received exception: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED]                                                                                                                                                              certificate verify failed (_ssl.c:590)>"
}

```

**Comment by caphrim007** [comment]
@selhorn it's not required. and its set to true by default because setting it to false by default would imply an insecure BIG-IP configuration. The reason that you're seeing that error is because it's functioning exactly how its supposed to.

You have no configured you SSL CA certificates correctly on your Ansible controller so that SSL verification of your remote BIG-IP can happen correctly. The error is telling you this; SSL verification failed.

So this is not a bug, its functioning as designed.

**Comment by selhorn** [comment]
OK thanks. Do you know why bigip_pool doesn't produce the same error?

**Comment by caphrim007** [comment]
it should. they use the same code

**Comment by selhorn** [comment]
Same machine, same setup. Here is playbook:
---

- name: Create a VIP, pool, pool members and nodes
  hosts: 10.10.10.10
  connection: local
  
  tasks:
      - name: Create a pool
        bigip_pool:
            server: "10.10.10.10"
            user: "admin"
            password: "blah"
            name: "web_pool"
            monitor_type: "and_list"
            monitors:
                - "http"
            lb_method: "ratio-member"

And here is output:
changed: [50.18.101.9] => {
    "changed": true,
    "invocation": {
        "module_args": {
            "description": null,
            "host": null,
            "lb_method": "ratio-member",
            "monitor_type": "and_list",
            "monitors": [
                "http"
            ],
            "name": "web_pool",
            "partition": "Common",
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "port": null,
            "quorum": null,
            "reselect_tries": null,
            "server": "50.18.101.9",
            "server_port": 443,
            "service_down_action": null,
            "slow_ramp_time": null,
            "state": "present",
            "user": "admin",
            "validate_certs": true
        }
    },
    "monitor": "/Common/http",
    "monitor_type": "and_list",
    "monitors": [
        "/Common/http"
    ],
    "partition": "Common"
}

You can take a look if you'd like--let me know.

**Comment by caphrim007** [comment]
the reason is because in 2.4 the virtual_server module still used the old soap methods of communicating with the device and bigip_pool does not. There was a bug in 2.4 that caused cert validation to be ignored, and it was being addressed for 2.5
